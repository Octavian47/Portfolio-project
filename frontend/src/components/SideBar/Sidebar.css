/* General Sidebar Styles */
.sidebar {
    width: 220px;
    background-color: #2e8b57;
    padding: 20px;
    position: fixed;
    height: 100%;
    top: 0;
    left: 0;
    overflow-y: auto;
    box-shadow: 2px 0 5px rgba(0, 0, 0, 0.1);
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    transition: transform 0.3s ease; /* Smooth transition for hiding and showing */
  }
  
  .sidebar ul {
    list-style: none;
    padding: 0;
    margin: 0;
    flex-grow: 1;
  }
  
  .sidebar li {
    margin: 10px 0;
  }
  
  .sidebar a {
    color: white;
    text-decoration: none;
    display: block;
    padding: 10px 15px;
    border-radius: 4px;
    font-weight: 500;
    font-size: 1rem;
    background-color: #4CAF50;
    transition: background-color 0.3s ease, transform 0.3s ease;
  }
  
  .sidebar a:hover {
    background-color: #45a049;
    transform: translateY(-2px);
  }
  
  .sidebar a:active {
    background-color: #3e8b47;
    transform: translateY(1px);
  }
  
  .sidebar .add-button {
    background-color: #4CAF50;
    text-align: center;
    margin-top: 20px;
    padding: 10px 20px;
  }
  
  .sidebar .add-button:hover {
    background-color: #45a049;
  }
  
  .sidebar .add-button:active {
    background-color: #3e8b47;
  }
  
  /* Media Queries for Responsiveness */
  @media (max-width: 768px) {
    .sidebar {
      transform: translateX(-100%); /* Hide the sidebar */
    }
  
    .sidebar.open {
      transform: translateX(0); /* Show the sidebar when open */
    }
  
    .hamburger-menu {
      display: block;
      position: fixed;
      top: 20px;
      left: 20px;
      z-index: 1000; /* Make sure it appears on top */
      background-color: #2e8b57;
      padding: 10px;
      border-radius: 4px;
      cursor: pointer;
    }
  
    .hamburger-menu div {
      width: 25px;
      height: 3px;
      background-color: white;
      margin: 5px 0;
      transition: 0.4s;
    }
  
    .hamburger-menu.open div:nth-child(1) {
      transform: rotate(-45deg) translate(-5px, 6px);
    }
  
    .hamburger-menu.open div:nth-child(2) {
      opacity: 0;
    }
  
    .hamburger-menu.open div:nth-child(3) {
      transform: rotate(45deg) translate(-5px, -6px);
    }
  }
  
  @media (min-width: 769px) {
    .hamburger-menu {
      display: none; /* Hide hamburger menu on larger screens */
    }
  }
  