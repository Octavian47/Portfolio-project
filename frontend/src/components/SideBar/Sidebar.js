import React from 'react';
import { Link } from 'react-router-dom';
import './Sidebar.css'; // Add styles for the sidebar

// Accept className and onClose as props
const Sidebar = ({ className, onClose }) => {
  return (
    <div className={`sidebar ${className}`}>
      <button className="close-btn" onClick={onClose}>
        &times;
      </button>
      <ul>
        <li>
          <Link to="/dashboard">List of Portfolio Items</Link>
        </li>
        <li>
          <Link to="/portfolio/add" className="add-button">Add Portfolio Item</Link>
        </li>
        <li>
          <a href="/" target="_blank" rel="noopener noreferrer">Live Portfolio</a>
        </li>
      </ul>
    </div>
  );
};

export default Sidebar;
