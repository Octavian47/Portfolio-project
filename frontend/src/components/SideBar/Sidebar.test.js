import React from 'react';
import { render, fireEvent, act } from '@testing-library/react'; // Import act to handle state updates
import Sidebar from './Sidebar';
import { BrowserRouter as Router } from 'react-router-dom'; // Router is needed to render links properly

describe('Sidebar Component', () => {
  // Test case: Ensure the Sidebar renders with all the expected links
  it('renders the Sidebar with all the expected links', () => {
    // Render the Sidebar component wrapped in a Router
    const { getByText } = render(
      <Router>
        <Sidebar className="open" onClose={jest.fn()} /> {/* Mock onClose for this test */}
      </Router>
    );

    // Assert that all expected links are in the document
    expect(getByText('List of Portfolio Items')).toBeInTheDocument();
    expect(getByText('Add Portfolio Item')).toBeInTheDocument();
    expect(getByText('Live Portfolio')).toBeInTheDocument();
  });

  // Test case: Ensure the onClose handler is called when the close button is clicked
  it('calls the onClose handler when the close button is clicked', () => {
    const handleClose = jest.fn(); // Create a mock function for onClose

    // Render the Sidebar with the mock onClose handler
    const { getByText } = render(
      <Router>
        <Sidebar className="open" onClose={handleClose} />
      </Router>
    );

    // Use act to ensure that any updates caused by the event are handled correctly
    act(() => {
      fireEvent.click(getByText('Ã—')); // Simulate a click on the close button
    });

    // Assert that the mock function was called exactly once
    expect(handleClose).toHaveBeenCalledTimes(1);
  });

  // Snapshot test: Ensure the Sidebar component renders consistently
  it('matches the snapshot', () => {
    // Render the Sidebar and create a snapshot
    const { asFragment } = render(
      <Router>
        <Sidebar className="open" onClose={jest.fn()} />
      </Router>
    );

    // Match the rendered output with the snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});
