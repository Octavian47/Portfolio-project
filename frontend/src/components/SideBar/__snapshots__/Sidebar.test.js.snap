// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sidebar open"
  >
    <button
      class="close-btn"
    >
      Ã—
    </button>
    <ul>
      <li>
        <a
          href="/dashboard"
        >
          List of Portfolio Items
        </a>
      </li>
      <li>
        <a
          class="add-button"
          href="/portfolio/add"
        >
          Add Portfolio Item
        </a>
      </li>
      <li>
        <a
          href="/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Live Portfolio
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
